packages/playground/hmr/hmr.ts

```diff
@@ -41,12 +41,41 @@ if (import.meta.hot) {
         update.type === 'css-update' && update.path.match('global.css')
     )
     if (cssUpdate) {
-      const el = document.querySelector('#global-css') as HTMLLinkElement
-      text('.css-prev', el.href)
-      // We don't have a vite:afterUpdate event, but updates are currently sync
-      setTimeout(() => {
-        text('.css-post', el.href)
-      }, 0)
+      text(
+        '.css-prev',
+        (document.querySelector('.global-css') as HTMLLinkElement).href
+      )
+
+      // We don't have a vite:afterUpdate event, but updates are currently
+      // sync. We need to wait until the tag has been swapped out, which
+      // includes the time taken to download and parse the new stylesheet.
+      let loop = 0
+      const updateAfterLinkTagHasBeenSwappedOut = () => {
+        loop = loop + 1
+
+        if (loop > 1000) {
+          return
+        }
+
+        const tags = document.querySelectorAll('.global-css')
+
+        // The order of these checks is important. We always want to check for
+        // 1 and return early.
+        if (tags.length === 1) {
+          text('.css-post', (tags[0] as HTMLLinkElement).href)
+          text('.ended-with-1-link-tag', 'yes')
+
+          return
+        }
+
+        if (tags.length === 2) {
+          text('.found-2-link-tags', 'yes')
+        }
+
+        setTimeout(updateAfterLinkTagHasBeenSwappedOut, 1)
+      }
+
+      setTimeout(updateAfterLinkTagHasBeenSwappedOut, 0)
```

Comment:
How about using `MutationObserver`?

Reply:
I've migrated this to mutation observer (TIL mutation observer! Silly back-end dev over here 👋)

Let me know if anything doesn't look right with it.